crates/wasm-encoder/src/component/types.rs

```diff
@@ -0,0 +1,574 @@
+const TABLE_DEFINITION_KIND: u8 = 0x03;
+const MEMORY_DEFINITION_KIND: u8 = 0x04;
+const GLOBAL_DEFINITION_KIND: u8 = 0x05;
+
+const ALIAS_TYPE_INSTANCE_EXPORT: u8 = 0x00;
+const ALIAS_TYPE_OUTER: u8 = 0x01;
+
+const OUTER_ALIAS_MODULE: u8 = 0x01;
+const OUTER_ALIAS_TYPE: u8 = 0x06;
+
+const INSTANCE_TYPEDEF_TYPE: u8 = 0x01;
+const INSTANCE_TYPEDEF_ALIAS: u8 = 0x05;
+const INSTANCE_TYPEDEF_EXPORT: u8 = 0x06;
+
+const MODULE_TYPEDEF_TYPE: u8 = 0x01;
+const MODULE_TYPEDEF_ALIAS: u8 = 0x05;
+const MODULE_TYPEDEF_EXPORT: u8 = 0x06;
+const MODULE_TYPEDEF_IMPORT: u8 = 0x02;
+
+/// Represents a definition kind.
+///
+/// This is used in aliases to specify the kind of the definition.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum DefinitionKind {
+    /// The definition is an instance.
+    Instance,
+    /// The definition is a module.
+    Module,
+    /// The definition is a function.
+    Function,
+    /// The definition is a table.
+    Table,
+    /// The definition is a memory.
+    Memory,
+    /// The definition is a global.
+    Global,
+}
+
+impl DefinitionKind {
+    pub(crate) fn encode(&self, bytes: &mut Vec<u8>) {
+        match self {
+            Self::Instance => bytes.push(INSTANCE_DEFINITION_KIND),
+            Self::Module => bytes.push(MODULE_DEFINITION_KIND),
+            Self::Function => bytes.push(FUNCTION_DEFINITION_KIND),
+            Self::Table => bytes.push(TABLE_DEFINITION_KIND),
+            Self::Memory => bytes.push(MEMORY_DEFINITION_KIND),
+            Self::Global => bytes.push(GLOBAL_DEFINITION_KIND),
+        }
+    }
+}
+
+/// Represents an index for an outer alias.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum OuterAliasIndex {
+    /// The index is a module index.
+    Module(u32),
+    /// The index is a type index.
+    Type(u32),
+}
+
+impl OuterAliasIndex {
+    pub(crate) fn encode(&self, bytes: &mut Vec<u8>) {
+        match self {
+            Self::Module(index) => {
+                bytes.extend(encoders::u32(*index));
+                bytes.push(OUTER_ALIAS_MODULE);
+            }
+            Self::Type(index) => {
+                bytes.extend(encoders::u32(*index));
+                bytes.push(OUTER_ALIAS_TYPE);
+            }
+        }
+    }
+}
+
+/// Represents an alias.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Alias<'a> {
+    /// The alias is to an instance export.
+    InstanceExport {
+        /// The index of the instance being aliased.
+        instance: u32,
+        /// The name of the export.
+        name: &'a str,
+        /// The kind of the definition being aliased.
+        kind: DefinitionKind,
+    },
+    /// The alias is to an index in an outer module.
+    Outer {
+        /// The count of the outer module.
+        ///
+        /// The count starts at 0 which denotes the current module.
+        count: u32,
+        /// The index from the outer module's index space being aliased.
+        index: OuterAliasIndex,
+    },
```

Comment:
Since these aliases are only in the type section I think only the "outer type" alias is needed here, right? Everything else would be invalid anyway?

Reply:
I'm not entirely sure I follow your question, but my understanding of the new format, and that could explain why there's no attempt to bolt the component's format on top of the core module format, is that sections may be defined in any order. The only validation of indexes is that, at the point they are read, the index exists (i.e. a prior section or the current section has defined it). Note that the outer alias count can also refer to the _current_ module at index 0, not just starting from the parent.  Thus types that refer to those indexes must be defined _after_ those referred-to indexes exist.

Reply:
Oh I mean that this `Alias`, which is distinct from aliases in the alias section, is only intended for pulling in outer types into an instance or module type. In that sense this for this specific `Alias` variant, which is only inteded for types, is always invalid for anything other than the "outer type" variant, right?

Reply:
Ah, I understand now. Yes, you are correct, the only expected case of an alias here is the outer form.

I've removed the `Alias` enum here with better builder patterns anyway, but I'll ensure that only outer aliases can be included in instance/module types.

Reply:
This should now be fixed (and the confusing `Alias` enum here removed entirely).